File: packages/sdk/src/clients/reputation.ts
import { type Account, type Address, type Hex, type PublicClient, type WalletClient, zeroHash } from 'viem';
import { reputationRegistryAbi } from '../abi/reputationRegistry.js';
import { WalletRequiredError } from '../errors.js';
import { FeedbackCategory, type GiveFeedbackParams } from '../types.js';
import { hashString, waitForReceipt } from '../utils.js';

export type ReputationClientOpts = {
  publicClient: PublicClient;
  walletClient?: WalletClient;
  account?: Account;
  reputationRegistry: Address;
};

export class ReputationClient {
  constructor(private readonly opts: ReputationClientOpts) {}

  /// Submit feedback for an agent. Per ERC-8004, an agent owner cannot submit
  /// feedback for their own agent (the contract reverts).
  async giveFeedback(params: GiveFeedbackParams) {
    const account = this.requireAccount('giveFeedback');
    const tag = params.tag ?? '';
    const metadataURI = params.metadataURI ?? '';
    const evidenceURI = params.evidenceURI ?? '';
    const comment = params.comment ?? '';
    const feedbackHash =
      params.feedbackHash ?? (comment ? hashString(comment) : (zeroHash as Hex));

    const hash = await this.opts.walletClient!.writeContract({
      account,
      chain: null,
      address: this.opts.reputationRegistry,
      abi: reputationRegistryAbi,
      functionName: 'giveFeedback',
      args: [
        BigInt(params.agentId),
        BigInt(params.score),
        params.feedbackType,
        tag,
        metadataURI,
        evidenceURI,
        comment,
        feedbackHash,
      ],
    });
    return waitForReceipt(this.opts.publicClient, hash);
  }

  /// Convenience: submit a positive (or negative) score with no metadata.
  /// Score is bounded by int128 range.
  async quickFeedback(
    agentId: bigint | number,
    score: number,
    category: FeedbackCategory = FeedbackCategory.Quality,
  ) {
    return this.giveFeedback({ agentId, score, feedbackType: category });
  }

  private requireAccount(op: string): Account {
    if (!this.opts.account || !this.opts.walletClient) {
      throw new WalletRequiredError(op);
    }
    return this.opts.account;
  }
}

File: packages/sdk-py/src/arckit/clients/reputation.py
"""ERC-8004 ReputationRegistry client."""

from __future__ import annotations

from web3 import Web3

from arckit._tx import TxContext
from arckit.abi import REPUTATION_REGISTRY_ABI
from arckit.types import FeedbackCategory
from arckit.utils import hash_string

ZERO_HASH = b"\x00" * 32


class ReputationClient:
    def __init__(self, ctx: TxContext, reputation_registry_address: str):
        self._ctx = ctx
        self._contract = ctx.w3.eth.contract(
            address=Web3.to_checksum_address(reputation_registry_address),
            abi=REPUTATION_REGISTRY_ABI,
        )

    def give_feedback(
        self,
        agent_id: int,
        score: int,
        feedback_type: FeedbackCategory = FeedbackCategory.QUALITY,
        tag: str = "",
        metadata_uri: str = "",
        evidence_uri: str = "",
        comment: str = "",
        feedback_hash: bytes | None = None,
    ) -> dict:
        """Submit feedback for an agent (per ERC-8004, owner cannot self-rate)."""
        if feedback_hash is None:
            feedback_hash = hash_string(comment) if comment else ZERO_HASH
        return self._ctx.send(
            self._contract.functions.giveFeedback(
                int(agent_id),
                int(score),
                int(feedback_type),
                tag,
                metadata_uri,
                evidence_uri,
                comment,
                feedback_hash,
            )
        )

    def quick_feedback(
        self,
        agent_id: int,
        score: int,
        category: FeedbackCategory = FeedbackCategory.QUALITY,
    ) -> dict:
        return self.give_feedback(agent_id=agent_id, score=score, feedback_type=category)
